### PR TITLE
Execute joshua_done script if it exists

### PR DESCRIPTION
**Pull request by sfc-gh-mpilman**
This can be used to upload more metadata to joshua after each run.

The current use-case for this is to accumulate code-probes in joshua

### CODE REVIEW

File: joshua/joshua_agent.py [hunk 1]
```diff
@@ -460,6 +494,7 @@ def _run_ensemble(
         properties = joshua_model.get_ensemble_properties(ensemble)
         compressed = properties.get("compressed", False)
         command = properties.get("test_command", "./joshua_test")
+        done_command = properties.get("end_command", "./joshua_done")
```
Comment (sfc-gh-anoyes) [inline joshua/joshua_agent.py:497]: Should we pick use `done_command` as the key in the db?

File: joshua/joshua_agent.py [hunk 2]
```diff
@@ -393,6 +394,39 @@ def cleanup(ensemble, where, seed, retcode=0, save_on="FAILURE", work_dir=None):
...
+
+    def cancel(self):
+        with self._lock:
+            self._m_cancelled = True
+
+    def _cancelled(self):
+        with self._lock:
+            return self._m_cancelled
+
+    def run(self, command, cwd, env, ensemble, seed, sanity):
```
Comment (sfc-gh-anoyes) [inline joshua/joshua_agent.py:410]: ensemble, seed, and sanity are not used

File: joshua/joshua_agent.py [hunk 3]
```diff
@@ -571,23 +606,38 @@ def _run_ensemble(
...
-                i = 0
-                while os.path.exists(to_write):
-                    to_write = os.path.join(where, "tmp", "console-{0}.log".format(i))
-                    i += 1
-
-                with open(to_write, "wb") as fout:
-                    fout.write(output)
+        del_output_file = not should_save(retcode, save_on)
+        to_write = os.path.join(where, "tmp", "console.log")
+        try:
```
Comment (sfc-gh-anoyes) [inline joshua/joshua_agent.py:611]: What's the motivation for unconditionally entering the block of code that was guarded by `should_save` previously?

**Review by sfc-gh-anoyes — COMMENTED** [review]
It'd be great if we could add a test for this

File: joshua/joshua_agent.py (hunk 3 above)
Comment (sfc-gh-mpilman) [reply joshua/joshua_agent.py:611]: We now need the file -- so `should_save` will determine whether the file will get deleted or not. But `joshua_done` will expect the file to be present while it is executed

File: joshua/joshua_agent.py (hunk 1 above)
Comment (sfc-gh-mpilman) [reply joshua/joshua_agent.py]: Sure, changed it

**Review by sfc-gh-anoyes — APPROVED** [review]

**Pull request merged**